### PR TITLE
Use null-terminated SerialNumber

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
#### Problem

Other strings (FirmwareRevisionString, ProductionRevisionString, ProductName, VendorName) are null-terminated. SerialNumber is not.

#### Change overview

* Aligned with FirmwareRevisionString, ProductionRevisionString, ProductName, VendorName
 which are also null-terminated
* Adapt SerialNumber test accordingly

#### Testing

Serial Number has a unit test.

Tested on nrf52840-dk lighting-app with
* CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER of maximum length
* serial number in ZephyrConfig
  * serial number of maximum length (32) stored with null-terminator cannot be loaded because of `bufSize - 1` in https://github.com/project-chip/connectedhomeip/blob/master/src/platform/Zephyr/ZephyrConfig.cpp#L179. this means that the stored value has to be without null-terminator. FYI: @Damian-Nordic

### CODE REVIEW

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 1]
```diff
@@ -79,22 +79,26 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
 }
 
 template <class ImplClass>
-CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetSerialNumber(char * buf, size_t bufSize, size_t & serialNumLen)
+CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetSerialNumber(char * buf, size_t bufSize)
 {
     CHIP_ERROR err;
-    err = Impl()->ReadConfigValueStr(ImplClass::kConfigKey_SerialNum, buf, bufSize, serialNumLen);
+    size_t serialNumLen = 0; // including null-terminator
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:85]: I'm not sure what this comment means.  At least the Linux implementation of `ReadConfigValueStr` puts the length of the string _not_ including the null terminator in `serialNumLen`.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 2]
```diff
@@ -79,22 +79,26 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
...
+    err                 = Impl()->ReadConfigValueStr(ImplClass::kConfigKey_SerialNum, buf, bufSize, serialNumLen);
+
+    ReturnErrorCodeIf(serialNumLen > bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:88]: This is assuming callee does not touch serialNumLen on error conditions.  Should this line move to after the CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND block and to after the check for failing err that should be there?

Also, if serialNumLen is length not including null terminator, then this check should be `>=`, not `>`.  We need to make sure we get this straight.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 3]
```diff
@@ -79,22 +79,26 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
...
+
 #ifdef CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
     if (CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER[0] != 0 && err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
     {
-        VerifyOrExit(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) <= bufSize, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        ReturnErrorCodeIf(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) > bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
         memcpy(buf, CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER, sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER));
-        serialNumLen = sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) - 1;
+        serialNumLen = sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER);
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:95]: This change seems wrong, because it will now include the null terminator in `serialNumLen`.....

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 4]
```diff
@@ -79,22 +79,26 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
...
         err          = CHIP_NO_ERROR;
     }
 #endif // CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
-    SuccessOrExit(err);
 
-exit:
+    ReturnErrorCodeIf(buf[serialNumLen - 1] != 0, CHIP_ERROR_INVALID_STRING_LENGTH);
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:100]: Again, this does not look like it will do the right check on Linux.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 5]
```diff
@@ -465,12 +469,11 @@ void GenericConfigurationManagerImpl<ImplClass>::LogDeviceConfig()
 {
     CHIP_ERROR err;
 
-    ChipLogProgress(DeviceLayer, "Device Configuration:");
+    ChipLogProgress(DeviceLayer, "Device Configuration (mab):");
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:472]: Why this change?

**Comment by markus-becker-tridonic-com** [comment]
@bzbarsky-apple please wait with review, I am still fixing up things.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 5 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp]: unintentionally committed. removed.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 3 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp]: fixed.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 4 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp]: fixed.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 1 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp]: fixed

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp]: fixed.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 6]
```diff
@@ -79,22 +79,25 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
...
+        ReturnErrorCodeIf(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) > bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
         memcpy(buf, CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER, sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER));
         serialNumLen = sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) - 1;
         err          = CHIP_NO_ERROR;
     }
 #endif // CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
-    SuccessOrExit(err);
 
-exit:
+    ReturnErrorCodeIf(serialNumLen >= bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```
Comment (bzbarsky-apple) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:98]: Needs `ReturnErrorOnFailure(err);` here, no?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp [hunk 7]
```diff
@@ -79,22 +79,25 @@ CHIP_ERROR GenericConfigurationManagerImpl<ImplClass>::GetFirmwareRevisionString
...
     if (CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER[0] != 0 && err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
     {
-        VerifyOrExit(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) <= bufSize, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        ReturnErrorCodeIf(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) > bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
         memcpy(buf, CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER, sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER));
         serialNumLen = sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) - 1;
         err          = CHIP_NO_ERROR;
     }
 #endif // CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
-    SuccessOrExit(err);
```
Comment (emargolis) [inline src/include/platform/internal/GenericConfigurationManagerImpl.cpp:95]: This check should stay, otherwise, the error returned in line 85 won't be reported.
Can be replaced by equivalent:
`ReturnErrorOnFailure(err);`

**Review by emargolis — APPROVED** [review]

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 6 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp:100]: Added before the two `ReturnErrorCodeIf`.

File: src/include/platform/internal/GenericConfigurationManagerImpl.cpp (hunk 7 above)
Comment (markus-becker-tridonic-com) [reply src/include/platform/internal/GenericConfigurationManagerImpl.cpp:95]: Added before the two `ReturnErrorCodeIf`.

**Review by jelderton — APPROVED** [review]
